packages/performance/src/services/transport_service.ts

```diff
@@ -73,60 +82,134 @@ function processQueue(timeOffset: number): void {
       return processQueue(DEFAULT_SEND_INTERVAL_MS);
     }
 
-    // Capture a snapshot of the queue and empty the "official queue".
-    const staged = [...queue];
-    queue = [];
-
-    /* eslint-disable camelcase */
-    // We will pass the JSON serialized event to the backend.
-    const log_event: Log[] = staged.map(evt => ({
-      source_extension_json_proto3: evt.message,
-      event_time_ms: String(evt.eventTime)
-    }));
-
-    const data: CcBatchLogFormat = {
-      request_time_ms: String(Date.now()),
-      client_info: {
-        client_type: 1, // 1 is JS
-        js_client_info: {}
-      },
-      log_source: SettingsService.getInstance().logSource,
-      log_event
-    };
-    /* eslint-enable camelcase */
-
-    fetch(SettingsService.getInstance().logEndPointUrl, {
-      method: 'POST',
-      body: JSON.stringify(data)
+    dispatchQueueEvents();
+  }, timeOffset);
+}
+
+function dispatchQueueEvents(): void {
+  // Capture a snapshot of the queue and empty the "official queue".
+  const staged = [...queue];
+  queue = [];
+
+  /* eslint-disable camelcase */
+  // We will pass the JSON serialized event to the backend.
+  const log_event: Log[] = staged.map(evt => ({
+    source_extension_json_proto3: evt.message,
+    event_time_ms: String(evt.eventTime)
+  }));
+
+  const data: TransportBatchLogFormat = {
+    request_time_ms: String(Date.now()),
+    client_info: {
+      client_type: 1, // 1 is JS
+      js_client_info: {}
+    },
+    log_source: SettingsService.getInstance().logSource,
+    log_event
+  };
+  /* eslint-enable camelcase */
+
+  postToEndpoint(data, staged).catch(() => {
+    /**
+     * If the request fails for some reason, add the events that were attempted
+     * back to the primary queue to retry later.
+     */
+    queue = [...staged, ...queue];
+    remainingTries--;
+    consoleLogger.info(`Tries left: ${remainingTries}.`);
+    processQueue(DEFAULT_SEND_INTERVAL_MS);
+  });
+}
+
+function postToEndpoint(
+  data: TransportBatchLogFormat,
+  staged: BatchEvent[]
+): Promise<void> {
+  // Gradually rollout traffic from cc to transport using remote config.
+  if (SettingsService.getInstance().shouldSendToTransport) {
+    return sendEventsToTransport(data, staged);
+  } else {
+    return sendEventsToCc(data);
+  }
+}
+
+function sendEventsToTransport(
```

Comment:
Other than the end point url, the rest of the sending  logic seems to be  reusable  between  cc and transport. Can  we merge them?